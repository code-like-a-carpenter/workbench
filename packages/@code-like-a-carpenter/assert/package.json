{
  "name": "@code-like-a-carpenter/assert",
  "description": "Wrapper around node's assert module which accepts a Producer instead of a string or already-constructed exception",
  "author": "Ian Remmel <1182361+ianwremmel@users.noreply.github.com> (https://www.ianwremmel.com)",
  "bugs": "https://www.github.com/code-like-a-carpenter/workbench/issues",
  "engines": {
    "node": "20.x"
  },
  "exports": {
    ".": {
      "import": {
        "types": "./dist/types/index.d.mts",
        "carpentry": "./src/index.ts",
        "default": "./dist/esm/index.mjs"
      },
      "require": {
        "types": "./dist/cjs-types/index.d.ts",
        "carpentry": "./src/index.ts",
        "default": "./dist/cjs/index.cjs"
      }
    },
    "./package.json": "./package.json"
  },
  "homepage": "https://www.github.com/code-like-a-carpenter/workbench/tree/main/packages/@code-like-a-carpenter/assert",
  "license": "MIT",
  "repository": "git@github.com:code-like-a-carpenter/workbench.git",
  "types": "dist/types",
  "publishConfig": {
    "access": "public"
  },
  "releaseAll": 5
}
