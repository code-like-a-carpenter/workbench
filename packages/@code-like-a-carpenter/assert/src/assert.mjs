// This package is a wrapper around node assert.
// eslint-disable-next-line no-restricted-imports
import nodeAssert, {AssertionError} from 'assert';

/**
 * @callback Provider
 * @returns {string | Error}
 */

/**
 * @param value {unknown}
 * @param message {string | Error | Provider}
 * @returns {asserts value}
 */
export function assert(value, message) {
  if (typeof message === 'string' || message instanceof Error) {
    nodeAssert.ok(value, message);
  } else {
    try {
      nodeAssert.ok(value);
    } catch (err) {
      const realError = message();
      if (realError instanceof Error) {
        throw realError;
      }
      throw new AssertionError({message: realError});
    }
  }
}

export {
  deepEqual,
  deepStrictEqual,
  // Removing the next line for now because it's missing from esbuild's node
  // builtins polyfill
  // doesNotMatch,
  doesNotReject,
  doesNotThrow,
  equal,
  fail,
  ifError,
  // Removing the next line for now because it's missing from esbuild's node
  // builtins polyfill
  // match,
  notDeepEqual,
  notDeepStrictEqual,
  notEqual,
  notStrictEqual,
  ok,
  rejects,
  strictEqual,
  throws,
  AssertionError,
} from 'assert';
