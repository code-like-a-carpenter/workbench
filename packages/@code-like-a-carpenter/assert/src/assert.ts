// This package is a wrapper around node assert.
// eslint-disable-next-line no-restricted-imports
import nodeAssert, {AssertionError} from 'assert';

type Provider = () => string | Error;

export function assert(
  value: unknown,
  message: string | Error | Provider
): asserts value {
  if (typeof message === 'string' || message instanceof Error) {
    nodeAssert.ok(value, message);
  } else {
    try {
      nodeAssert.ok(value);
    } catch (err) {
      const realError = message();
      if (realError instanceof Error) {
        throw realError;
      }
      throw new AssertionError({message: realError});
    }
  }
}

export {
  deepEqual,
  deepStrictEqual,
  doesNotMatch,
  doesNotReject,
  doesNotThrow,
  equal,
  fail,
  ifError,
  match,
  notDeepEqual,
  notDeepStrictEqual,
  notEqual,
  notStrictEqual,
  ok,
  rejects,
  strictEqual,
  throws,
  AssertionError,
} from 'assert';
