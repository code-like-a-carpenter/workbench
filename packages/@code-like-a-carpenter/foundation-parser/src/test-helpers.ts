import {buildSchema} from 'graphql';

import {schema as coreSchema} from '@code-like-a-carpenter/foundation-intermediate-representation';

import {ParserConfigSchema} from './config.ts';
import {parse} from './parser.ts';

export async function parseSchema(raw: string) {
  const schema = buildSchema(`${coreSchema}\n${raw}`);

  return parse(
    schema,
    [],
    ParserConfigSchema.parse({
      actionsModuleId: './placeholder/actions.ts',
      dependenciesModuleId: './placeholder/dependencies.ts',
      requireExtensions: true,
    }),
    {
      outputFile: './placeholder/output',
    }
  );
}
