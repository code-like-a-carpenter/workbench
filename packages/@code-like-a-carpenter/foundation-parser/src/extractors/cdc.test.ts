import assert from 'node:assert';

import {parseSchema} from '../test-helpers';

describe('@reacts', () => {
  it('configures an event handler for the model', async () => {
    const schema = /* GraphQL */ `
      """
      A summary of all the accounts on a particular plan.
      """
      type PlanMetric implements Model & Timestamped & Versioned
        # This is a little weird looking, but it's basically a way to allow scans
        # without generally building it in as a feature. This strategy should only be
        # in very small tables where it won't matter that every record is in a single
        # partition. In this case, we're talking about all the plans a startup might
        # offer times 2 for each boolean (cancelled and onFreeTrial), so it's probably
        # less than 20 records. If you find yourself with many more records, you
        # should look at using something other than DynamoDB for aggregating this
        # data.
        @compositeKey(
          pkFields: []
          pkPrefix: "BUSINESS_METRIC"
          skFields: ["onFreeTrial", "planName"]
          skPrefix: "PLAN"
        )
        @triggers(
          event: "UPSERT"
          handler: "../src/react--plan-metric--upsert"
        ) {
        createdAt: Date!
        id: ID!
        updatedAt: Date!
        version: Int!
        cancelled: Boolean!
        count: Int!
        monthlyRecurringRevenueInCents: Int!
        onFreeTrial: Boolean!
        planName: String!
      }
    `;

    const parsed = await parseSchema(schema);

    const [table] = parsed.tables;
    assert(table);
    expect(table.hasCdc).toBe(true);
    assert(table.hasCdc);
    expect(table.dispatcherConfig).toMatchInlineSnapshot(`
      {
        "memorySize": 384,
        "timeout": 90,
      }
    `);

    const [model] = parsed.models;
    assert(model);
    expect(model.changeDataCaptureConfig).toHaveLength(1);
    expect(model.changeDataCaptureConfig).toMatchInlineSnapshot(`
      [
        {
          "dispatcherConfig": {
            "memorySize": 384,
            "timeout": 90,
          },
          "event": "UPSERT",
          "handlerConfig": {
            "memorySize": 256,
            "timeout": 30,
          },
          "handlerModuleId": "../src/react--plan-metric--upsert",
          "readableTables": [],
          "sourceModelName": "PlanMetric",
          "type": "TRIGGER",
          "writableTables": [],
        },
      ]
    `);

    expect(parsed).toMatchSnapshot();
  });
});

describe('@enriches', () => {
  it('configures the handler to update one model when another changes', async () => {
    const schema = /* GraphQL */ `
      """
      A customer account.
      """
      type Account implements Model & Versioned & Timestamped
        @compositeKey(
          pkFields: ["externalId"]
          pkPrefix: "ACCOUNT"
          skFields: []
        ) {
        createdAt: Date!
        id: ID!
        updatedAt: Date!
        version: Int!
        cancelled: Boolean!
        externalId: String!
        planName: String
      }
      """
      An event describing a change to a customer's subscription status.
      """
      type SubscriptionEvent implements Model & Timestamped & Versioned
        @enriches(
          event: "UPSERT"
          handler: "../src/enrich--subscription--upsert--account"
          targetModel: "Account"
        )
        @ledger
        @compositeKey(
          pkFields: ["externalId"]
          pkPrefix: "ACCOUNT"
          skFields: ["effectiveDate"]
          skPrefix: "SUBSCRIPTION_EVENT"
        ) {
        createdAt: Date!
        id: ID!
        updatedAt: Date!
        version: Int!
        cancelled: Boolean!
        effectiveDate: Date!
        externalId: String!
        monthlyPriceInCents: Int!
        onFreeTrial: Boolean!
        planName: String!
      }
    `;

    const parsed = await parseSchema(schema);

    const table = parsed.tables.find(
      ({tableName}) => tableName === 'TableSubscriptionEvent'
    );
    assert(table);
    expect(table.hasCdc).toBe(true);
    assert(table.hasCdc);
    expect(table.dispatcherConfig).toMatchInlineSnapshot(`
      {
        "memorySize": 384,
        "timeout": 90,
      }
    `);

    const model = parsed.models.find(
      ({typeName}) => typeName === 'SubscriptionEvent'
    );
    assert(model);
    expect(model.changeDataCaptureConfig).toHaveLength(1);
    expect(model.changeDataCaptureConfig).toMatchInlineSnapshot(`
      [
        {
          "dispatcherConfig": {
            "memorySize": 384,
            "timeout": 90,
          },
          "event": "UPSERT",
          "handlerConfig": {
            "memorySize": 256,
            "timeout": 30,
          },
          "handlerModuleId": "../src/enrich--subscription--upsert--account",
          "sourceModelName": "SubscriptionEvent",
          "targetModelName": "Account",
          "type": "ENRICHER",
          "writableTables": [
            "TableAccount",
          ],
        },
      ]
    `);

    expect(parsed).toMatchSnapshot();
  });
});
