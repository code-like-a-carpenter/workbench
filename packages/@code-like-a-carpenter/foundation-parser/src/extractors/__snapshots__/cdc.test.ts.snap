// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@enriches configures the handler to update one model when another changes 3`] = `
{
  "additionalImports": [],
  "dependenciesModuleId": "dependencies.ts",
  "models": [
    {
      "changeDataCaptureConfig": [],
      "consistent": false,
      "enablePointInTimeRecovery": true,
      "fields": [
        {
          "columnName": "_ct",
          "columnNamesForRead": [
            "_ct",
          ],
          "computeFunction": undefined,
          "ean": ":createdAt",
          "eav": "#createdAt",
          "fieldName": "createdAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "id",
          "columnNamesForRead": [
            "id",
            "id",
          ],
          "computeFunction": undefined,
          "ean": ":id",
          "eav": "#id",
          "fieldName": "id",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "ID",
        },
        {
          "columnName": "_md",
          "columnNamesForRead": [
            "_md",
          ],
          "computeFunction": undefined,
          "ean": ":updatedAt",
          "eav": "#updatedAt",
          "fieldName": "updatedAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "_v",
          "columnNamesForRead": [
            "_v",
          ],
          "computeFunction": undefined,
          "ean": ":version",
          "eav": "#version",
          "fieldName": "version",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Int",
        },
        {
          "columnName": "cancelled",
          "columnNamesForRead": [
            "cancelled",
            "cancelled",
          ],
          "computeFunction": undefined,
          "ean": ":cancelled",
          "eav": "#cancelled",
          "fieldName": "cancelled",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Boolean",
        },
        {
          "columnName": "external_id",
          "columnNamesForRead": [
            "external_id",
            "externalId",
          ],
          "computeFunction": undefined,
          "ean": ":externalId",
          "eav": "#externalId",
          "fieldName": "externalId",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "String",
        },
        {
          "columnName": "plan_name",
          "columnNamesForRead": [
            "plan_name",
            "planName",
          ],
          "computeFunction": undefined,
          "ean": ":planName",
          "eav": "#planName",
          "fieldName": "planName",
          "isDateType": false,
          "isRequired": false,
          "isScalarType": true,
          "typeName": "String",
        },
      ],
      "isLedger": false,
      "isPublic": false,
      "primaryKey": {
        "isComposite": true,
        "partitionKeyFields": [
          {
            "columnName": "external_id",
            "columnNamesForRead": [
              "external_id",
              "externalId",
            ],
            "computeFunction": undefined,
            "ean": ":externalId",
            "eav": "#externalId",
            "fieldName": "externalId",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "String",
          },
        ],
        "partitionKeyIsSingleField": true,
        "partitionKeyName": "pk",
        "partitionKeyPrefix": "ACCOUNT",
        "sortKeyFields": [],
        "sortKeyIsSingleField": false,
        "sortKeyName": "sk",
        "sortKeyPrefix": "",
        "type": "primary",
      },
      "secondaryIndexes": [],
      "table": {
        "dependenciesModuleId": "../dependencies.ts",
        "enablePointInTimeRecovery": true,
        "enableStreaming": false,
        "hasCdc": false,
        "hasPublicModels": false,
        "hasTtl": false,
        "primaryKey": {
          "isComposite": true,
          "partitionKeyFields": [
            {
              "columnName": "external_id",
              "columnNamesForRead": [
                "external_id",
                "externalId",
              ],
              "computeFunction": undefined,
              "ean": ":externalId",
              "eav": "#externalId",
              "fieldName": "externalId",
              "isDateType": false,
              "isRequired": true,
              "isScalarType": true,
              "typeName": "String",
            },
          ],
          "partitionKeyIsSingleField": true,
          "partitionKeyName": "pk",
          "partitionKeyPrefix": "ACCOUNT",
          "sortKeyFields": [],
          "sortKeyIsSingleField": false,
          "sortKeyName": "sk",
          "sortKeyPrefix": "",
          "type": "primary",
        },
        "secondaryIndexes": [],
        "tableName": "TableAccount",
      },
      "ttlConfig": undefined,
      "typeName": "Account",
    },
    {
      "changeDataCaptureConfig": [
        {
          "actionsModuleId": "../actions.ts",
          "directory": "placeholder/enricher--subscription-event--upsert--account",
          "event": "UPSERT",
          "filename": "enricher--subscription-event--upsert--account",
          "functionName": "FnEnricherSEUpsertA5b59364d",
          "handlerImportName": "Placeholder",
          "handlerModuleId": "../../../src/enrich--subscription--upsert--account.ts",
          "memorySize": 256,
          "nestedStackLocation": "./cdc.yml",
          "readableTables": [],
          "runtimeModuleId": "@code-like-a-carpenter/foundation-runtime",
          "sourceModelName": "SubscriptionEvent",
          "targetModelName": "Account",
          "timeout": 30,
          "type": "ENRICHER",
          "writableTables": [
            "TableAccount",
          ],
        },
      ],
      "consistent": false,
      "enablePointInTimeRecovery": true,
      "fields": [
        {
          "columnName": "_ct",
          "columnNamesForRead": [
            "_ct",
          ],
          "computeFunction": undefined,
          "ean": ":createdAt",
          "eav": "#createdAt",
          "fieldName": "createdAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "id",
          "columnNamesForRead": [
            "id",
            "id",
          ],
          "computeFunction": undefined,
          "ean": ":id",
          "eav": "#id",
          "fieldName": "id",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "ID",
        },
        {
          "columnName": "_md",
          "columnNamesForRead": [
            "_md",
          ],
          "computeFunction": undefined,
          "ean": ":updatedAt",
          "eav": "#updatedAt",
          "fieldName": "updatedAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "_v",
          "columnNamesForRead": [
            "_v",
          ],
          "computeFunction": undefined,
          "ean": ":version",
          "eav": "#version",
          "fieldName": "version",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Int",
        },
        {
          "columnName": "cancelled",
          "columnNamesForRead": [
            "cancelled",
            "cancelled",
          ],
          "computeFunction": undefined,
          "ean": ":cancelled",
          "eav": "#cancelled",
          "fieldName": "cancelled",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Boolean",
        },
        {
          "columnName": "effective_date",
          "columnNamesForRead": [
            "effective_date",
            "effectiveDate",
          ],
          "computeFunction": undefined,
          "ean": ":effectiveDate",
          "eav": "#effectiveDate",
          "fieldName": "effectiveDate",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "external_id",
          "columnNamesForRead": [
            "external_id",
            "externalId",
          ],
          "computeFunction": undefined,
          "ean": ":externalId",
          "eav": "#externalId",
          "fieldName": "externalId",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "String",
        },
        {
          "columnName": "monthly_price_in_cents",
          "columnNamesForRead": [
            "monthly_price_in_cents",
            "monthlyPriceInCents",
          ],
          "computeFunction": undefined,
          "ean": ":monthlyPriceInCents",
          "eav": "#monthlyPriceInCents",
          "fieldName": "monthlyPriceInCents",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Int",
        },
        {
          "columnName": "on_free_trial",
          "columnNamesForRead": [
            "on_free_trial",
            "onFreeTrial",
          ],
          "computeFunction": undefined,
          "ean": ":onFreeTrial",
          "eav": "#onFreeTrial",
          "fieldName": "onFreeTrial",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Boolean",
        },
        {
          "columnName": "plan_name",
          "columnNamesForRead": [
            "plan_name",
            "planName",
          ],
          "computeFunction": undefined,
          "ean": ":planName",
          "eav": "#planName",
          "fieldName": "planName",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "String",
        },
      ],
      "isLedger": true,
      "isPublic": false,
      "primaryKey": {
        "isComposite": true,
        "partitionKeyFields": [
          {
            "columnName": "external_id",
            "columnNamesForRead": [
              "external_id",
              "externalId",
            ],
            "computeFunction": undefined,
            "ean": ":externalId",
            "eav": "#externalId",
            "fieldName": "externalId",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "String",
          },
        ],
        "partitionKeyIsSingleField": true,
        "partitionKeyName": "pk",
        "partitionKeyPrefix": "ACCOUNT",
        "sortKeyFields": [
          {
            "columnName": "effective_date",
            "columnNamesForRead": [
              "effective_date",
              "effectiveDate",
            ],
            "computeFunction": undefined,
            "ean": ":effectiveDate",
            "eav": "#effectiveDate",
            "fieldName": "effectiveDate",
            "isDateType": true,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "Date",
          },
        ],
        "sortKeyIsSingleField": true,
        "sortKeyName": "sk",
        "sortKeyPrefix": "SUBSCRIPTION_EVENT",
        "type": "primary",
      },
      "secondaryIndexes": [],
      "table": {
        "dependenciesModuleId": "../dependencies.ts",
        "dispatcherConfig": {
          "batchSize": 10,
          "dependenciesModuleId": "../dependencies.ts",
          "directory": "placeholder/dispatcher-table-subscription-event",
          "filename": "dispatcher-table-subscription-event",
          "functionName": "TableSubscriptionEventCDCDispatcher",
          "maximumRetryAttempts": 3,
          "memorySize": 256,
          "nestedStackLocation": "./dispatcher.yml",
          "runtimeModuleId": "@code-like-a-carpenter/foundation-runtime",
          "timeout": 30,
        },
        "enablePointInTimeRecovery": true,
        "enableStreaming": true,
        "hasCdc": true,
        "hasPublicModels": false,
        "hasTtl": false,
        "primaryKey": {
          "isComposite": true,
          "partitionKeyFields": [
            {
              "columnName": "external_id",
              "columnNamesForRead": [
                "external_id",
                "externalId",
              ],
              "computeFunction": undefined,
              "ean": ":externalId",
              "eav": "#externalId",
              "fieldName": "externalId",
              "isDateType": false,
              "isRequired": true,
              "isScalarType": true,
              "typeName": "String",
            },
          ],
          "partitionKeyIsSingleField": true,
          "partitionKeyName": "pk",
          "partitionKeyPrefix": "ACCOUNT",
          "sortKeyFields": [
            {
              "columnName": "effective_date",
              "columnNamesForRead": [
                "effective_date",
                "effectiveDate",
              ],
              "computeFunction": undefined,
              "ean": ":effectiveDate",
              "eav": "#effectiveDate",
              "fieldName": "effectiveDate",
              "isDateType": true,
              "isRequired": true,
              "isScalarType": true,
              "typeName": "Date",
            },
          ],
          "sortKeyIsSingleField": true,
          "sortKeyName": "sk",
          "sortKeyPrefix": "SUBSCRIPTION_EVENT",
          "type": "primary",
        },
        "secondaryIndexes": [],
        "tableName": "TableSubscriptionEvent",
      },
      "ttlConfig": undefined,
      "typeName": "SubscriptionEvent",
    },
  ],
  "tables": [
    {
      "dependenciesModuleId": "../dependencies.ts",
      "enablePointInTimeRecovery": true,
      "enableStreaming": false,
      "hasCdc": false,
      "hasPublicModels": false,
      "hasTtl": false,
      "primaryKey": {
        "isComposite": true,
        "partitionKeyFields": [
          {
            "columnName": "external_id",
            "columnNamesForRead": [
              "external_id",
              "externalId",
            ],
            "computeFunction": undefined,
            "ean": ":externalId",
            "eav": "#externalId",
            "fieldName": "externalId",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "String",
          },
        ],
        "partitionKeyIsSingleField": true,
        "partitionKeyName": "pk",
        "partitionKeyPrefix": "ACCOUNT",
        "sortKeyFields": [],
        "sortKeyIsSingleField": false,
        "sortKeyName": "sk",
        "sortKeyPrefix": "",
        "type": "primary",
      },
      "secondaryIndexes": [],
      "tableName": "TableAccount",
    },
    {
      "dependenciesModuleId": "../dependencies.ts",
      "dispatcherConfig": {
        "batchSize": 10,
        "dependenciesModuleId": "../dependencies.ts",
        "directory": "placeholder/dispatcher-table-subscription-event",
        "filename": "dispatcher-table-subscription-event",
        "functionName": "TableSubscriptionEventCDCDispatcher",
        "maximumRetryAttempts": 3,
        "memorySize": 256,
        "nestedStackLocation": "./dispatcher.yml",
        "runtimeModuleId": "@code-like-a-carpenter/foundation-runtime",
        "timeout": 30,
      },
      "enablePointInTimeRecovery": true,
      "enableStreaming": true,
      "hasCdc": true,
      "hasPublicModels": false,
      "hasTtl": false,
      "primaryKey": {
        "isComposite": true,
        "partitionKeyFields": [
          {
            "columnName": "external_id",
            "columnNamesForRead": [
              "external_id",
              "externalId",
            ],
            "computeFunction": undefined,
            "ean": ":externalId",
            "eav": "#externalId",
            "fieldName": "externalId",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "String",
          },
        ],
        "partitionKeyIsSingleField": true,
        "partitionKeyName": "pk",
        "partitionKeyPrefix": "ACCOUNT",
        "sortKeyFields": [
          {
            "columnName": "effective_date",
            "columnNamesForRead": [
              "effective_date",
              "effectiveDate",
            ],
            "computeFunction": undefined,
            "ean": ":effectiveDate",
            "eav": "#effectiveDate",
            "fieldName": "effectiveDate",
            "isDateType": true,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "Date",
          },
        ],
        "sortKeyIsSingleField": true,
        "sortKeyName": "sk",
        "sortKeyPrefix": "SUBSCRIPTION_EVENT",
        "type": "primary",
      },
      "secondaryIndexes": [],
      "tableName": "TableSubscriptionEvent",
    },
  ],
}
`;

exports[`@reacts configures an event handler for the model 3`] = `
{
  "additionalImports": [],
  "dependenciesModuleId": "dependencies.ts",
  "models": [
    {
      "changeDataCaptureConfig": [
        {
          "actionsModuleId": "../actions.ts",
          "directory": "placeholder/react--plan-metric--upsert",
          "event": "UPSERT",
          "filename": "react--plan-metric--upsert",
          "functionName": "FnReactPMUpsert2b2e8d7d",
          "handlerImportName": "Placeholder",
          "handlerModuleId": "../../../src/react--plan-metric--upsert.ts",
          "memorySize": 256,
          "nestedStackLocation": "./cdc.yml",
          "readableTables": [],
          "runtimeModuleId": "@code-like-a-carpenter/foundation-runtime",
          "sourceModelName": "PlanMetric",
          "timeout": 30,
          "type": "REACTOR",
          "writableTables": [],
        },
      ],
      "consistent": false,
      "enablePointInTimeRecovery": true,
      "fields": [
        {
          "columnName": "_ct",
          "columnNamesForRead": [
            "_ct",
          ],
          "computeFunction": undefined,
          "ean": ":createdAt",
          "eav": "#createdAt",
          "fieldName": "createdAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "id",
          "columnNamesForRead": [
            "id",
            "id",
          ],
          "computeFunction": undefined,
          "ean": ":id",
          "eav": "#id",
          "fieldName": "id",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "ID",
        },
        {
          "columnName": "_md",
          "columnNamesForRead": [
            "_md",
          ],
          "computeFunction": undefined,
          "ean": ":updatedAt",
          "eav": "#updatedAt",
          "fieldName": "updatedAt",
          "isDateType": true,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Date",
        },
        {
          "columnName": "_v",
          "columnNamesForRead": [
            "_v",
          ],
          "computeFunction": undefined,
          "ean": ":version",
          "eav": "#version",
          "fieldName": "version",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Int",
        },
        {
          "columnName": "cancelled",
          "columnNamesForRead": [
            "cancelled",
            "cancelled",
          ],
          "computeFunction": undefined,
          "ean": ":cancelled",
          "eav": "#cancelled",
          "fieldName": "cancelled",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Boolean",
        },
        {
          "columnName": "count",
          "columnNamesForRead": [
            "count",
            "count",
          ],
          "computeFunction": undefined,
          "ean": ":count",
          "eav": "#count",
          "fieldName": "count",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Int",
        },
        {
          "columnName": "monthly_recurring_revenue_in_cents",
          "columnNamesForRead": [
            "monthly_recurring_revenue_in_cents",
            "monthlyRecurringRevenueInCents",
          ],
          "computeFunction": undefined,
          "ean": ":monthlyRecurringRevenueInCents",
          "eav": "#monthlyRecurringRevenueInCents",
          "fieldName": "monthlyRecurringRevenueInCents",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Int",
        },
        {
          "columnName": "on_free_trial",
          "columnNamesForRead": [
            "on_free_trial",
            "onFreeTrial",
          ],
          "computeFunction": undefined,
          "ean": ":onFreeTrial",
          "eav": "#onFreeTrial",
          "fieldName": "onFreeTrial",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "Boolean",
        },
        {
          "columnName": "plan_name",
          "columnNamesForRead": [
            "plan_name",
            "planName",
          ],
          "computeFunction": undefined,
          "ean": ":planName",
          "eav": "#planName",
          "fieldName": "planName",
          "isDateType": false,
          "isRequired": true,
          "isScalarType": true,
          "typeName": "String",
        },
      ],
      "isLedger": false,
      "isPublic": false,
      "primaryKey": {
        "isComposite": true,
        "partitionKeyFields": [],
        "partitionKeyIsSingleField": false,
        "partitionKeyName": "pk",
        "partitionKeyPrefix": "BUSINESS_METRIC",
        "sortKeyFields": [
          {
            "columnName": "on_free_trial",
            "columnNamesForRead": [
              "on_free_trial",
              "onFreeTrial",
            ],
            "computeFunction": undefined,
            "ean": ":onFreeTrial",
            "eav": "#onFreeTrial",
            "fieldName": "onFreeTrial",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "Boolean",
          },
          {
            "columnName": "plan_name",
            "columnNamesForRead": [
              "plan_name",
              "planName",
            ],
            "computeFunction": undefined,
            "ean": ":planName",
            "eav": "#planName",
            "fieldName": "planName",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "String",
          },
        ],
        "sortKeyIsSingleField": false,
        "sortKeyName": "sk",
        "sortKeyPrefix": "PLAN",
        "type": "primary",
      },
      "secondaryIndexes": [],
      "table": {
        "dependenciesModuleId": "../dependencies.ts",
        "dispatcherConfig": {
          "batchSize": 10,
          "dependenciesModuleId": "../dependencies.ts",
          "directory": "placeholder/dispatcher-table-plan-metric",
          "filename": "dispatcher-table-plan-metric",
          "functionName": "TablePlanMetricCDCDispatcher",
          "maximumRetryAttempts": 3,
          "memorySize": 256,
          "nestedStackLocation": "./dispatcher.yml",
          "runtimeModuleId": "@code-like-a-carpenter/foundation-runtime",
          "timeout": 30,
        },
        "enablePointInTimeRecovery": true,
        "enableStreaming": true,
        "hasCdc": true,
        "hasPublicModels": false,
        "hasTtl": false,
        "primaryKey": {
          "isComposite": true,
          "partitionKeyFields": [],
          "partitionKeyIsSingleField": false,
          "partitionKeyName": "pk",
          "partitionKeyPrefix": "BUSINESS_METRIC",
          "sortKeyFields": [
            {
              "columnName": "on_free_trial",
              "columnNamesForRead": [
                "on_free_trial",
                "onFreeTrial",
              ],
              "computeFunction": undefined,
              "ean": ":onFreeTrial",
              "eav": "#onFreeTrial",
              "fieldName": "onFreeTrial",
              "isDateType": false,
              "isRequired": true,
              "isScalarType": true,
              "typeName": "Boolean",
            },
            {
              "columnName": "plan_name",
              "columnNamesForRead": [
                "plan_name",
                "planName",
              ],
              "computeFunction": undefined,
              "ean": ":planName",
              "eav": "#planName",
              "fieldName": "planName",
              "isDateType": false,
              "isRequired": true,
              "isScalarType": true,
              "typeName": "String",
            },
          ],
          "sortKeyIsSingleField": false,
          "sortKeyName": "sk",
          "sortKeyPrefix": "PLAN",
          "type": "primary",
        },
        "secondaryIndexes": [],
        "tableName": "TablePlanMetric",
      },
      "ttlConfig": undefined,
      "typeName": "PlanMetric",
    },
  ],
  "tables": [
    {
      "dependenciesModuleId": "../dependencies.ts",
      "dispatcherConfig": {
        "batchSize": 10,
        "dependenciesModuleId": "../dependencies.ts",
        "directory": "placeholder/dispatcher-table-plan-metric",
        "filename": "dispatcher-table-plan-metric",
        "functionName": "TablePlanMetricCDCDispatcher",
        "maximumRetryAttempts": 3,
        "memorySize": 256,
        "nestedStackLocation": "./dispatcher.yml",
        "runtimeModuleId": "@code-like-a-carpenter/foundation-runtime",
        "timeout": 30,
      },
      "enablePointInTimeRecovery": true,
      "enableStreaming": true,
      "hasCdc": true,
      "hasPublicModels": false,
      "hasTtl": false,
      "primaryKey": {
        "isComposite": true,
        "partitionKeyFields": [],
        "partitionKeyIsSingleField": false,
        "partitionKeyName": "pk",
        "partitionKeyPrefix": "BUSINESS_METRIC",
        "sortKeyFields": [
          {
            "columnName": "on_free_trial",
            "columnNamesForRead": [
              "on_free_trial",
              "onFreeTrial",
            ],
            "computeFunction": undefined,
            "ean": ":onFreeTrial",
            "eav": "#onFreeTrial",
            "fieldName": "onFreeTrial",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "Boolean",
          },
          {
            "columnName": "plan_name",
            "columnNamesForRead": [
              "plan_name",
              "planName",
            ],
            "computeFunction": undefined,
            "ean": ":planName",
            "eav": "#planName",
            "fieldName": "planName",
            "isDateType": false,
            "isRequired": true,
            "isScalarType": true,
            "typeName": "String",
          },
        ],
        "sortKeyIsSingleField": false,
        "sortKeyName": "sk",
        "sortKeyPrefix": "PLAN",
        "type": "primary",
      },
      "secondaryIndexes": [],
      "tableName": "TablePlanMetric",
    },
  ],
}
`;
